tree_detection_framework/detection/detector.py

```
@@ -421,6 +422,184 @@ def save_model(self, save_file: PATH_TYPE):
+        for pred in batch_preds:
+
+            # Get the Instances object
+            segmentations = [dic["segmentation"].astype(float) for dic in pred]
+
+            # Convert each mask to a shapely multipolygon
+            shapely_objects = [
+                mask_to_shapely(pred_mask) for pred_mask in segmentations
+            ]
+
+            all_geometries.append(shapely_objects)
+
+            # Get prediction scores
+            scores = [dic["stability_score"] for dic in pred]
+            # unlabelled masks
+            labels = None
+            all_data_dicts.append({"score": scores, "labels": labels})
+
+        return all_geometries, all_data_dicts
+
+
+class SAMV2Detector(Detector):
+    def __init__(self):
+        self.device = (
+            torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+        )
+
+    def setup_predictor(self):
+        sam2_checkpoint = "checkpoints/sam2.1_hiera_large.pt"
+        model_cfg = "configs/sam2.1/sam2.1_hiera_l.yaml"
+        # Build the SAM2 model
+        self.sam2 = build_sam2(
+            model_cfg, sam2_checkpoint, device=self.device, apply_postprocessing=False
+        )
+
+        # Create the automatic mask generator
+        self.mask_generator = SAM2AutomaticMaskGenerator(self.sam2)
+
+    def call_predict(self, batch):
+        """
+        Args:
+            batch (Tensor): 4 dims Tensor with the first dimension having number of images in the batch
+
+        Returns:
+            masks List[List[Dict]]: list of dictionaries for each mask in the batch
+        """
+        self.setup_predictor()
+
+        with torch.no_grad():
+            masks = []
+            for original_image in batch:
+                original_image = (
+                    original_image.permute(1, 2, 0).byte().numpy()
+                )  # (4, h, w)
+                # permuted_image = original_image.permute(1, 2, 0)
+                rgb_image = original_image[:, :, :3]  # (h, w, 3)
```

Subsetting to 3 channels is completely fine, but you could also raise an error if you get a different number of channels. Up to you. As an aside, would this support one-channel data? We have been using one-channel canopy height rasters in this framework as well.

---

based on [this](https://github.com/facebookresearch/segment-anything/issues/642) issue it seems it only supports 3 channels. in that case is it possible to turn 1 channel to 3 channel (grayscale -> rgb?). i tried condensing the original_image to 1 channel with `original_image[:, :, :1]` and it seemed to error out so im not too sure if its supported but you can give it a try with the canopy height rasters

---

Ok thanks for looking into this